pkg/cmd/join/exec.go

```diff
@@ -213,84 +255,129 @@ func (o *Options) validate() error {
 
-	files := []string{}
-	// If Deployment/klusterlet is not deployed, deploy it
-	if !available {
-		files = append(files,
-			"join/klusterlets.crd.yaml",
-			"join/namespace.yaml",
-			"join/service_account.yaml",
-			"join/cluster_role.yaml",
-			"join/cluster_role_binding.yaml",
-		)
-	}
-	files = append(files,
-		"join/namespace_agent.yaml",
-		"join/bootstrap_hub_kubeconfig.yaml",
-	)
+	if o.mode == InstallModeControlplaneAgent {
+		_, err = kubeClient.CoreV1().Namespaces().Get(context.TODO(), "multicluster-controlplane-agent", metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				_, err = kubeClient.CoreV1().Namespaces().Create(context.TODO(), &corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "multicluster-controlplane-agent",
+					},
+				}, metav1.CreateOptions{})
+				if err != nil {
+					return err
+				}
+			} else {
+				return err
+			}
+		}
+
+		files := []string{}
 
-	if o.mode == InstallModeHosted {
 		files = append(files,
-			"join/hosted/external_managed_kubeconfig.yaml",
+			"controlplane/bootstrap_hub_kubeconfig.yaml",
+			"controlplane/clusterrole.yaml",
+			"controlplane/clusterrolebinding-admin.yaml",
+			"controlplane/clusterrolebinding.yaml",
+			"controlplane/role.yaml",
+			"controlplane/rolebinding.yaml",
+			"controlplane/serviceaccount.yaml",
+			"controlplane/deployment.yaml",
 		)
-	}
 
-	r := reader.NewResourceReader(o.builder, o.ClusteradmFlags.DryRun, o.Streams)
-	err = r.Apply(scenario.Files, o.values, files...)
-	if err != nil {
-		return err
-	}
-
-	if !available {
-		err = r.Apply(scenario.Files, o.values, "join/operator.yaml")
+		err = r.Apply(scenario.Files, o.values, files...)
 		if err != nil {
 			return err
 		}
-	}
 
-	if !o.ClusteradmFlags.DryRun {
-		if err := wait.WaitUntilCRDReady(apiExtensionsClient, "klusterlets.operator.open-cluster-management.io", o.wait); err != nil {
+		if o.wait && !o.ClusteradmFlags.DryRun {
+			err = waitUntilControlplaneAgentConditionIsTrue(o.ClusteradmFlags.KubectlFactory, int64(o.ClusteradmFlags.Timeout), "multicluster-controlplane-agent")
+			if err != nil {
+				return err
+			}
+		}
+
+	} else {
+		available, err := checkIfRegistrationOperatorAvailable(o.ClusteradmFlags.KubectlFactory)
+		if err != nil {
 			return err
 		}
-	}
 
-	err = r.Apply(scenario.Files, o.values, "join/klusterlets.cr.yaml")
-	if err != nil {
-		return err
-	}
+		files := []string{}
+		// If Deployment/klusterlet is not deployed, deploy it
+		if !available {
+			files = append(files,
+				"join/klusterlets.crd.yaml",
+				"join/namespace.yaml",
+				"join/service_account.yaml",
+				"join/cluster_role.yaml",
+				"join/cluster_role_binding.yaml",
+			)
+		}
+		files = append(files,
+			"join/namespace_agent.yaml",
+			"join/bootstrap_hub_kubeconfig.yaml",
+		)
 
-	klusterletNamespace := o.values.Klusterlet.KlusterletNamespace
-	agentNamespace := o.values.Klusterlet.AgentNamespace
+		if o.mode == InstallModeHosted {
```

Comment:
hosted mode should be independent with using standalone agent or not, I think it should be possible that use hosted standalone agent, right?